announce mirage 3.5.0 (#625)
mirage/mirage-www: tmpl/blog/announcing-mirage-35-release.md

On the OCaml tooling side, since MirageOS 3.0.0 we did several major changes:
- moved most packages to [dune](https://dune.build/) (formerly jbuilder) and began using [dune-release](https://github.com/samoht/dune-release) for smooth developer experience and simple releases,
- require [opam](https://opam.ocaml.org) to be version 2.0.2 or later, allowing `pin-depends` in `config.ml`. `pin-depends` allows you to depend on a development branch of any opam package for your unikernel,
- adjusted documentation to adhere to [odoc](https://github.com/ocaml/odoc/) requirements,
- the `mirage` command-line utility now emits lower and upper bounds of opam packages, allowing uncompromising deprecation of packages,
- support for OCaml 4.06.0 (and above), where `safe-string` is enabled by default. Strings are immutable now!!,
- remove usage of `result` package, which has incorporated into `Pervasives` since OCaml 4.03.0.

> We are working on further changes (namely revising the `mirage` internal build system, at the moment using `ocamlbuild`, `ocamlfind`, `pkg-config`, and `make`) to make MirageOS more developer-friendly by using [dune](https://dune.build). Changing the internal build tooling allows for monorepos, incremental builds, cross-compilation, and other oft-requested improvements.

Reviewer: This one's forward looking and so a bit odd in the release announcement for 3.5.0 -- perhaps remove it until the mirage 4 release? Or shift it towards the end in a 'next steps' section

This is similar to the Ethernet signature - it writes the IPv4 header and sends a packet. It also supports fragmentation (including setting the do-not-fragment bit for path MTU discovery) -- whenever the payload is too big for a single frame, it is sent as multiple fragmented IPv4 packets. Additionally, setting the time-to-live is now supported, meaning we now can implement traceroute!
The API used to include two functions, `allocate_frame` and `write`, where only buffers allocated by the former should be used in the latter. This has been combined into a single function that takes a fill function and a list of payloads. This change is for maximum flexibility: a higher layer can either construct its header and payload, and pass it to `write` as payload argument (the `buffer list`), which is then copied into the buffer(s) allocated by the network device, or the upper layer can provide the callback `fill` function to assemble its data into the buffer allocated by the network device, to avoid copying. Of course, both can be used - the outgoing packet contains the IPv4 header, and possibly the buffer until the offset returned by `fill`, and afterwards the payload.

> 

Reviewer: This is all epic :-)

Author: oh, and I should describe the fragmentation and reassembly... oops

- [dns 1.0.0](https://github.com/mirage/ocaml-dns), see [this announcement](https://discuss.ocaml.org/t/ann-major-releases-of-cohttp-conduit-dns-tcpip)
- [conduit 1.0.0](https://github.com/mirage/ocaml-conduit), see [this announcement](https://discuss.ocaml.org/t/ann-major-releases-of-cohttp-conduit-dns-tcpip)

## More features and bugfixes

- A [httpaf device](https://github.com/mirage/mirage/pull/955) is now part of mirage
- [libvirt.xml is generated for virtio target](https://github.com/mirage/mirage/pull/903)
- [Unix target now include -tags thread](https://github.com/mirage/mirage/issues/861) (for mirage-framebuffer SDL support)
- Various modules (IPv6, DHCP) are explicit about their dependency to the random device
- [QubesDB can be requested in config.ml when the target is Xen](https://github.com/mirage/mirage/pull/807)

You may also want to read the [MirageOS 3.2.0 announcement](https://discuss.ocaml.org/t/ann-mirage-3-2-0) and the [MirageOS 3.3.0 announcement](https://discuss.ocaml.org/t/mirage-3-3-0-released).

## Next steps

> We are working on [further](https://github.com/mirage/mirage/issues/969) [changes](https://github.com/mirage/functoria/pull/167) which revise the `mirage` internal build system to [dune](https://dune.build). At the moment it uses `ocamlbuild`, `ocamlfind`, `pkg-config`, and `make`. The goal of this change is to make MirageOS more developer-friendly. On the horizon we have MirageOS unikernel monorepos, incremental builds, pain-free cross-compilation, documentation generation, ...

Reviewer 2: ... and an Irmin 2.0 release :-)

Author: I pushed 3e96dc4 (ok @samoht? @mato?). we can also just get rid of the next steps section since we don't want to promise too much, but may as well be a nice way to keep people interested. :)

Major API improvements (in [this PR](https://github.com/mirage/mirage-kv/pull/14), since 2.0.0):
- The `key` is now a path (list of segments) instead of a `string`
- The `value` type is now a `string`
- The new function `list : t -> key -> (string * [`Value|`Dictionary], error) result io` was added
- The function `get : t -> key -> (value, error) result io` is now provided (used to be named `read` and requiring an `offset` and `length` parameter)
- The functions `last_modified : t -> key -> (int * int64, error) result io` and `digest : t -> key -> (string, error) result io` have been introduced
- The function `size` was removed.
- The signature `RW` for read-write key-value stores extends `RO` with three functions `set`, `remove`, and `batch`

> There is now a [non-persistent in-memory implementation](https://github.com/mirage/mirage-kv-mem) of a read-write key-value store available. Other implementations (such as [crunch](https://github.com/mirage/ocaml-crunch), [mirage-kv-unix](https://github.com/mirage/mirage-kv-unix), [mirage-fs](https://github.com/mirage/mirage-fs), [tar](https://github.com/mirage/ocaml-tar) have been adapted, as well as cliens of mirage-kv (dns, cohttp, tls)).

Reviewer 3: ```suggestion
There is now a [non-persistent in-memory implementation](https://github.com/mirage/mirage-kv-mem) of a read-write key-value store available. Other implementations (such as [crunch](https://github.com/mirage/ocaml-crunch), [mirage-kv-unix](https://github.com/mirage/mirage-kv-unix), [mirage-fs](https://github.com/mirage/mirage-fs), [tar](https://github.com/mirage/ocaml-tar) have been adapted, as well as clients of mirage-kv (dns, cohttp, tls)).
```

### [TCP/IP](https://github.com/mirage/mirage-tcpip)

> The IPv4 implementation now has support for [fragment reassembly](https://github.com/mirage/mirage-tcpip/pull/375). Each incoming IPv4 fragment is checked for the more fragment and offset fields. If these are non-zero, the fragment is processed by the [fragment cache](https://mirage.github.io/mirage-tcpip/tcpip/Fragments/index.html), which uses a [least recently used](https://github.com/pqwy/lru) data structure of maximum size 256kB content shared by all incoming fragments. If there is any overlap in fragments, the entire packet is dropped ([avoiding security issues](https://eprint.iacr.org/2015/1020.pdf)). Fragments may arrive out of order. The code is [heavily unit-tested](https://github.com/mirage/mirage-tcpip/blob/v3.7.1/test/test_ipv4.ml#L49-L203). Each IPv4 packet may at most be in 16 fragments (to minimise CPU DoS with lots of small fragments), the timeout between the first and last fragment is 10 seconds.

Reviewer 3: ```suggestion
The IPv4 implementation now has support for [fragment reassembly](https://github.com/mirage/mirage-tcpip/pull/375). Each incoming IPv4 fragment is checked for the "more fragments" and "offset" fields. If these are non-zero, the fragment is processed by the [fragment cache](https://mirage.github.io/mirage-tcpip/tcpip/Fragments/index.html), which uses a [least recently used](https://github.com/pqwy/lru) data structure of maximum size 256kB content shared by all incoming fragments. If there is any overlap in fragments, the entire packet is dropped ([avoiding security issues](https://eprint.iacr.org/2015/1020.pdf)). Fragments may arrive out of order. The code is [heavily unit-tested](https://github.com/mirage/mirage-tcpip/blob/v3.7.1/test/test_ipv4.ml#L49-L203). Each IPv4 packet may at most be in 16 fragments (to minimise CPU DoS with lots of small fragments), the timeout between the first and last fragment is 10 seconds.
```

Reviewer 3: (the diff is not very clear, but this adds quotes around `"more fragments"`)

The API used to include two functions, `allocate_frame` and `write`, where only buffers allocated by the former should be used in the latter. This has been combined into a single function that takes a fill function and a list of payloads. This change is for maximum flexibility: a higher layer can either construct its header and payload, and pass it to `write` as payload argument (the `buffer list`), which is then copied into the buffer(s) allocated by the network device, or the upper layer can provide the callback `fill` function to assemble its data into the buffer allocated by the network device, to avoid copying. Of course, both can be used - the outgoing packet contains the IPv4 header, and possibly the buffer until the offset returned by `fill`, and afterwards the payload.

The TCP implementation has [preliminary keepalive support](https://github.com/mirage/mirage-tcpip/pull/338).

### [Solo5](https://github.com/solo5/solo5)

- MirageOS 3.0.0 used the 0.2.0 release of solo5
- The `ukvm` target was renamed to `hvt`, where `solo5-hvt` is the monitoring process

> - Support for [FreeBSD bhyve](http://bhyve.org/) and [OpenBSD VMM](https://man.openbsd.org/vmm.4) hypervisor (within the hvt target)

Reviewer: These new features seem like a really big deal for users, and hidden away somewhat.  We could trumpet "new platform support" for FreeBSD and OpenBSD and Linux/arm64 support right at the top?

Author: feel free to directly push to this branch!

- [base64 3.0.0](https://github.com/mirage/base64) performance and API changes, see [this announcement](https://discuss.ocaml.org/t/ann-major-release-of-base64-article)
- [git 2.0.0](https://github.com/mirage/ocaml-git), read [this announcement](https://discuss.ocaml.org/t/ann-ocaml-git-2-0), as well as [its design and implementation](https://discuss.ocaml.org/t/ocaml-git-git-design-and-implementation)
- [io-page 2.0.0](https://github.com/mirage/io-page), see [this announcement](https://discuss.ocaml.org/t/ann-io-page-2-0-0-with-packaging-changes)
- [cohttp 2.0.0](https://github.com/mirage/ocaml-cohttp), see [this announcement](https://discuss.ocaml.org/t/ann-major-releases-of-cohttp-conduit-dns-tcpip)
- [dns 1.0.0](https://github.com/mirage/ocaml-dns), see [this announcement](https://discuss.ocaml.org/t/ann-major-releases-of-cohttp-conduit-dns-tcpip)
- [conduit 1.0.0](https://github.com/mirage/ocaml-conduit), see [this announcement](https://discuss.ocaml.org/t/ann-major-releases-of-cohttp-conduit-dns-tcpip)

## More features and bugfixes

> - A [httpaf device](https://github.com/mirage/mirage/pull/955) is now part of mirage

Reviewer: ```suggestion
- More HTTP server choices are supported via a new [httpaf device](https://github.com/mirage/mirage/pull/955) that permits the [high performance httpaf](https://github.com/inhabitedtype/httpaf) stack to run as a unikernel now.
```
